src/communication/xcWebSocketPublisher.js

```diff
@@ -9,29 +9,62 @@ define(function () {
         var codes = this.configuration.getCodes(componentName, stateMachineName);
-        var publish = this.configuration.getPublisherDetails(codes.componentCode, codes.stateMachineCode);
-
+        var eventWithoutStateMachineRef = this.getEventWithoutStateMachineRef(codes.componentCode, codes.stateMachineCode);
         var event = {
-            "Header": {
-                "StateMachineCode": { "Case": "Some", "Fields": [parseInt(codes.stateMachineCode)] },
-                "ComponentCode": { "Case": "Some", "Fields": [parseInt(codes.componentCode)] },
-                "EventCode": parseInt(publish.eventCode),
-                "IncomingType": 0,
-                "MessageType": { "Case": "Some", "Fields": [publish.messageType] }
-            },
+            "Header": eventWithoutStateMachineRef.header,
             "JsonMessage": JSON.stringify(jsonMessage)
         };
         return {
             event: event,
-            routingKey: publish.routingKey
+            routingKey: eventWithoutStateMachineRef.routingKey
+        };
+    }
+
+
+    Publisher.prototype.getEventWithoutStateMachineRef = function (componentCode, stateMachineCode) {
+        var publisher = this.configuration.getPublisherDetails(componentCode, stateMachineCode);
+        var header = {
+            "StateMachineCode": { "Case": "Some", "Fields": [parseInt(stateMachineCode)] },
+            "ComponentCode": { "Case": "Some", "Fields": [parseInt(componentCode)] },
+            "EventCode": parseInt(publisher.eventCode),
+            "IncomingType": 0,
+            "MessageType": { "Case": "Some", "Fields": [publisher.messageType] }
         };
+        return {
+            header: header,
+            routingKey: publisher.routingKey
+        }
     }
 
 
     Publisher.prototype.send = function(componentName, stateMachineName, jsonMessage) {
         var data = this.getEventToSend(componentName, stateMachineName, jsonMessage);
         this.webSocket.send(convertToWebsocketInputFormat(data));
     }
+    
+
+    Publisher.prototype.getEventToSendUsingStateMachineRef = function (stateMachineRef, jsonMessage) {
+        var componentCode = stateMachineRef.ComponentCode.Fields[0];
+        var stateMachineCode = stateMachineRef.StateMachineCode.Fields[0];
+        var eventWithoutStateMachineRef = this.getEventWithoutStateMachineRef(componentCode, stateMachineCode);
+        var headerStateMachineRef = {
+            "AgentId": stateMachineRef.AgentId,
+            "StateMachineId": stateMachineRef.StateMachineId,
+        };
+        return {
+            event: {
+                "Header": mergeJsonObjects(headerStateMachineRef, eventWithoutStateMachineRef.header),
+                "JsonMessage": JSON.stringify(jsonMessage)
+            },
+            routingKey: eventWithoutStateMachineRef.routingKey
+        };
+    }
+
+
+    Publisher.prototype.sendStatemachineRef = function (stateMachineRef, jsonMessage) {
```

Comment:
Hard to understand the real purpose of this function
sendMessageWithStatemachineRef perhaps ?
I'm thinking to another cool thing (for api users):
adding a send method to statemachienref could be intuitive. no ?